include/core/Ref.hpp

```diff
@@ -21,7 +21,7 @@ class Ref {
 
 		unref();
 
-		reference = p_from.reference;
+		reference = Object::cast_to<T>(p_from.reference);
```

Comment:
The cast makes sense, but looks like the location is not right. `ref` here expects a reference where `T` is the same between the source and destination references (not even sure why it gets called while `InputEventScreenTouch` and `InputEventMouseButton` are incompatible?).
I would have expected this cast to occur in a function that expects a different `T`.

Reply:
Ok the real place to put the cast should actually be the places with `// TODO We need a safe cast`.
The reason `ref` ends up assigning the pointer is because of a cast that happened already, but it was done without any check.

Reply:
Ok.
Can I close this pull request and let you fix this issue?

Reply:
Yeah, I'm working on a fix, though having a bit of trouble enforcing T being a Reference at compile-time... might be able to fix the cast but to have the extra nicety it might need more work